Python/Product/LiveShare/PythonLanguageServiceProviderCallback.cs

```diff
@@ -55,89 +53,117 @@ internal class PythonLanguageServiceProviderCallback : ILanguageServiceProviderC
+            // via the intermediate JSON format which is backwards compatible.
+            switch (method.Name) {
+                case Methods.InitializeName:
+                    return Initialize<TOut>();
+                case Methods.TextDocumentCompletionName:
+                case Methods.TextDocumentHoverName:
+                case Methods.TextDocumentSignatureHelpName:
+                case Methods.TextDocumentDefinitionName:
+                case Methods.TextDocumentReferencesName:
+                    return await DispatchToLanguageServer(method, param, cancellationToken);
+                default:
+                    return default(TOut);
+            }
+        }
+
+        private async Task<PythonLanguageClient> FindClientAsync(Uri uri) {
+            if (uri == null) {
                 return null;
             }
 
-            if (!_analyzerCache.TryGetValue(document.Uri, out var analyzer)) {
-                var filePath = document.Uri.LocalPath;
-                if (string.IsNullOrEmpty(filePath)) {
-                    return null;
+            if (!_clientCache.TryGetValue(uri, out var client)) {
+                client = await GetOrCreateClientAsync(uri);
+                if (client != null) {
+                    client = _clientCache.GetOrAdd(uri, client);
                 }
+            }
 
-                if (_uiThread != null) {
-                    // TODO: Use URI for more accurate lookup
-                    analyzer = await _uiThread.InvokeTask(async () =>
-                        (await _serviceProvider.FindAllAnalyzersForFile(filePath)).FirstOrDefault() as VsProjectAnalyzer
-                    );
-                }
+            return client;
+        }
 
-                analyzer = _analyzerCache.GetOrAdd(document.Uri, analyzer);
+        private async Task<PythonLanguageClient> GetOrCreateClientAsync(Uri uri) {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            var filePath = uri.LocalPath;
+            if (string.IsNullOrEmpty(filePath)) {
+                return null;
             }
 
-            return analyzer;
+            var contentTypeProvider = _serviceProvider.GetComponentModel().GetService<PyFilePathToContentTypeProvider>();
+            if (contentTypeProvider.TryGetContentTypeForFilePath(filePath, out var contentType)) {
+                return PythonLanguageClient.FindLanguageClient(contentType.TypeName);
+            }
+
+            return null;
         }
 
         /// <summary>
         /// Helper function for tests, enabling this class to be tested without needing
         /// a service provider or UI thread.
         /// </summary>
-        internal void SetAnalyzer(Uri documentUri, VsProjectAnalyzer analyzer) {
-            _analyzerCache[documentUri] = analyzer;
+        internal void SetClient(Uri documentUri, PythonLanguageClient client) {
+            _clientCache[documentUri] = client;
```

Comment:
Consider adding a check for null client here.  I think SetClient can be called with results from FindLanguageClient with might be null

Reply:
It's technically valid, all the code in this class can handle a null entry in the dictionary. We could also want to add tests later that specifically test that behavior when it's null, so I think not throwing is fine.